extensions/mongodb/src/main/java/com/hazelcast/jet/mongodb/impl/WriteMongoP.java

```diff
@@ -0,0 +1,592 @@
+                getOutbox(),
+                context,
+                context.processingGuarantee(),
+                () -> new MongoTransactionId(context, processorIndex),
+                txId -> new MongoTransaction(txId, logger),
+                txId -> {
+                    MongoTransaction mongoTransaction = activeTransactions.get(txId);
+                    refreshTransaction(mongoTransaction, true);
+                },
+                () -> {
+                    for (MongoTransaction tx : activeTransactions.values()) {
+                        refreshTransaction(tx, false);
+                    }
+                }
+        );
+    }
+
+    @Override
+    public boolean tryProcessWatermark(@Nonnull Watermark watermark) {
+        return true;
+    }
+
+    @Override
+    protected boolean tryProcess(int ordinal, @Nonnull Object item) {
+        return transactionUtility.tryProcess();
+    }
+
+    @Override
+    public void process(int ordinal, @Nonnull Inbox inbox) {
+        if (!connection.reconnectIfNecessary()) {
+            return;
+        }
+
+        try {
+            MongoTransaction mongoTransaction = transactionUtility.activeTransaction();
+
+            ArrayList<IN> items = drainItems(inbox);
+            @SuppressWarnings("DataFlowIssue")
+            Map<MongoCollectionKey, List<I>> itemsPerCollection = items
+                    .stream()
+                    .map(intermediateMappingFn)
+                    .map(e -> tuple2(collectionPicker.pick(e), e))
+                    .collect(groupingBy(Tuple2::f0, mapping(Tuple2::getValue, toList())));
+
+            for (Map.Entry<MongoCollectionKey, List<I>> entry : itemsPerCollection.entrySet()) {
+                MongoCollectionKey collectionKey = entry.getKey();
+                MongoCollection<I> collection = collectionKey.get(connection.client(), documentType);
+
+                List<WriteModel<I>> writes = new ArrayList<>();
+
+                for (I item : entry.getValue()) {
+                    WriteModel<I> write = writeModelFn.apply(item);
+                    writes.add(write);
+                }
+                if (transactionUtility.usesTransactionLifecycle()) {
+                    checkNotNull(mongoTransaction, "there is no active transaction");
+                    mongoTransaction.addWrites(collectionKey, writes);
+                } else {
+                    collection.bulkWrite(writes);
+                }
+            }
+
+            for (int i = 0; i < items.size(); i++) {
+                inbox.remove();
+            }
+        } catch (MongoBulkWriteException e) {
+            throw new JetException(e);
+        } catch (Exception e) {
+            logger.info("Unable to process Mongo Sink: " + e.getMessage());
+            // not removing from inbox, so it will be retried
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private static <I> ArrayList<I> drainItems(Inbox inbox) {
```

Comment:
Inbox class has drainItems() method. May we can use that one ?

Reply:
I want to sub-batch writes, so that transaction size is not too big